CRM/Upgrade/Incremental/MessageTemplates.php

```diff
@@ -0,0 +1,144 @@
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2018
+ */
+class CRM_Upgrade_Incremental_MessageTemplates {
+
+  /**
+   * Version we are upgrading to.
+   *
+   * @var string
+   */
+  protected $upgradeVersion;
+
+  /**
+   * @return string
+   */
+  public function getUpgradeVersion() {
+    return $this->upgradeVersion;
+  }
+
+  /**
+   * @param string $upgradeVersion
+   */
+  public function setUpgradeVersion($upgradeVersion) {
+    $this->upgradeVersion = $upgradeVersion;
+  }
+
+  /**
+   * CRM_Upgrade_Incremental_MessageTemplates constructor.
+   *
+   * @param string $upgradeVersion
+   */
+  public function __construct($upgradeVersion) {
+    $this->setUpgradeVersion($upgradeVersion);
+  }
+
+  /**
+   * Get any templates that have been updated.
+   *
+   * @return array
+   */
+  protected function getTemplateUpdates() {
+    return [
+      [
+        'version' => '5.4.alpha1',
+        'upgrade_descriptor' => ts('Use email greeting at top where available'),
+        'templates' => [
+          ['name' => 'membership_online_receipt', 'type' => 'text'],
+          ['name' => 'membership_online_receipt', 'type' => 'html'],
+        ]
+      ],
+    ];
+  }
+
+  /**
+   * Get any required template updates.
+   *
+   * @return array
+   */
+  public function getTemplatesToUpdate() {
+    $templates = $this->getTemplateUpdates();
+    $return = [];
+    foreach ($templates as $templateArray) {
+      if ($templateArray['version'] === $this->getUpgradeVersion()) {
+        foreach ($templateArray['templates'] as $template) {
+          $return[$template['name'] . '_' . $template['type']] = $template;
+        }
+      }
+    }
+    return $return;
+  }
+
+  /**
+   * Get the upgrade messages.
+   */
+  public function getUpgradeMessages() {
+    $updates = $this->getTemplatesToUpdate();
+    $messages = [];
+    foreach ($updates as $key => $value) {
+      $messages[$key] = $value['upgrade_descriptor'];
+    }
```

Comment:
This will need tweaking since 'upgrade_descriptor' is not in the 'templates' sub-arrays returned now by `getTemplatesToUpdate()`

Reply:
Good point - have tweaked & added a test